### PR TITLE
index_add decomposed into index_put through TM Tensor

### PR DESCRIPTION
**Pull request by lorenzbc**
Combined all the commits from the index-add-to-put_TMTensor branch and move it to a new branch that's rebased

### CODE REVIEW

**Comment by ramiro050** [comment]
You can run the lit tests locally by doing `cmake --build build --target check-torch-mlir`

**Comment by lorenzbc** [comment]
@ramiro050 not sure why but the actual python unit test is failing. What could cause a "failed to legalize operation 'torch.aten.index_add' that was explicitly marked illegal" error?

**Comment by ramiro050** [comment]
> @ramiro050 not sure why but the actual python unit test is failing. What could cause a "failed to legalize operation 'torch.aten.index_add' that was explicitly marked illegal" error?

That error means that your conversion pattern failed, so the op `torch.aten.index_add` stayed in the IR.

From the error message, you can tell that the input does have a rank, so the failure has to be in the `getTensorRank(updates) != inputRank` check. I think this is because the unit test is not right. In your unit test, the input has rank 3 and the `source` tensor has rank 2 according to the annotation.

File: python/torch_mlir_e2e_test/test_suite/index_put.py [hunk 1]
```diff
@@ -818,3 +818,27 @@ def forward(self, input, index, value):
...
+      ([-1, -1], torch.float32, True),
+  ])
+  def forward(self, input, dim, index, value):
+    return torch.ops.aten.index_add(input, dim, index, value)
+
+
+@register_test_case(module_factory=lambda: IndexAddModule())
+def IndexAddModule_basic(module, tu: TestUtils):
+  module.forward(torch.rand((3, 5, 2), dtype=torch.float32),
+          torch.tensor(0.0),
```
Comment (ramiro050) [inline python/torch_mlir_e2e_test/test_suite/index_put.py:842]: This is a zero-rank tensor, so the annotation should be `[]`.

File: python/torch_mlir_e2e_test/test_suite/index_put.py [hunk 2]
```diff
@@ -818,3 +818,27 @@ def forward(self, input, index, value):
...
+          torch.tensor([0, 1]),
```
Comment (ramiro050) [inline python/torch_mlir_e2e_test/test_suite/index_put.py:843]: This is an int tensor, which does not match the dtype in the annotation

File: python/torch_mlir_e2e_test/test_suite/index_put.py [hunk 3]
```diff
@@ -818,3 +818,27 @@ def forward(self, input, index, value):
...
+          torch.rand((2, )))
```
Comment (ramiro050) [inline python/torch_mlir_e2e_test/test_suite/index_put.py:844]: This is a rank-1 tensor, which does not match the rank-2 annotation

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: python/torch_mlir_e2e_test/test_suite/index_put.py [hunk 4]
```diff
@@ -818,3 +818,27 @@ def forward(self, input, index, value):
...
+  @export
+  @annotate_args([
+      None,
+      ([-1, -1, -1], torch.float32, True),
+      ([-1], torch.float32, True),
+      ([-1], torch.float32, True),
+      ([-1, -1], torch.float32, True),
+  ])
+  def forward(self, input, dim, index, value):
+    return torch.ops.aten.index_add(input, dim, index, value)
```
Comment (ramiro050) [inline python/torch_mlir_e2e_test/test_suite/index_put.py:836]: can you also add an `alpha` value?

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: python/torch_mlir_e2e_test/test_suite/index_put.py (hunk 2 above)
Comment (lorenzbc) [reply python/torch_mlir_e2e_test/test_suite/index_put.py:843]: Can you clarify what you mean here?

File: python/torch_mlir_e2e_test/test_suite/index_put.py [hunk 5]
```diff
@@ -818,3 +818,27 @@ def forward(self, input, index, value):
...
+  def __init__(self):
+    super().__init__()
+
+  @export
+  @annotate_args([
+      None,
+      ([-1, -1, -1], torch.float32, True),
+      ([-1], torch.float32, True),
+      ([-1], torch.float32, True),
+      ([-1, -1], torch.float32, True),
```
Comment (vivekkhandelwal1) [inline python/torch_mlir_e2e_test/test_suite/index_put.py:833]: These annotations should match with the dtype and shape of the arguments passed in `module.forward()`

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

File: python/torch_mlir_e2e_test/test_suite/index_put.py (hunk 2 above)
Comment (ramiro050) [reply python/torch_mlir_e2e_test/test_suite/index_put.py:843]: The tensor has `dtype` int:

```python
In [3]: torch.tensor([0, 1]).dtype
Out[3]: torch.int64
```

which does not match the corresponding annotation here:

https://github.com/llvm/torch-mlir/blob/a8e50a68720ab37e832cc121224d5a8fbb8a78b9/python/torch_mlir_e2e_test/test_suite/index_put.py#L832

**Comment by lorenzbc** [comment]
The op is still marked as illegal despite the changes even though now the annotations match the types and shapes of the arguments

**Comment by ramiro050** [comment]
> The op is still marked as illegal despite the changes even though now the annotations match the types and shapes of the arguments

The test has inputs that are invalid in the decomposition. Here,

https://github.com/llvm/torch-mlir/blob/0c9084f7e14bb1479c7d20088d076b5eae3827a1/lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp#L1659

there is an expectation for the `source` tensor to have the same rank as the `input` tensor, but in your test the `source` tensor has rank 1, while the `input` tensor has rank 3

https://github.com/llvm/torch-mlir/blob/0c9084f7e14bb1479c7d20088d076b5eae3827a1/python/torch_mlir_e2e_test/test_suite/index_put.py#L830-L833

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 6]
```diff
@@ -1637,6 +1637,56 @@ class DecomposeAtenIndexPutOp : public OpRewritePattern<AtenIndexPutOp> {
...
+        op.getLoc(), expandUpdatesType, updates, torchCstOne);
+
+    Value indicesAsTensorList = rewriter.create<PrimListConstructOp>(
+        op.getLoc(), Torch::ListType::get(updates.getType()),
+        Value(op.index()));
+
+    Value accumulateTrue = rewriter.create<Torch::ConstantBoolOp>(op.getLoc(), true);
+    Value cstFalse = rewriter.create<Torch::ConstantBoolOp>(op.getLoc(), false);
+
+    rewriter.replaceOpWithNewOp<ValsemVariantAtenIndexPutImplOp>(
```
Comment (navahgar) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1682]: This decomposition seems to ignore the `dim` parameter of `torch.index_add`. AFAICT, there is no way to include that in `index_put`. So, how is this going to work when `dim != 0`? Am I missing something?

Comment (navahgar) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1682]: But I don't see the indices being transformed to the form that you mention here. I believe that should be done using the dim value and I don't see the dim being used anywhere here.

Comment (ramiro050) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1682]: This PR is still a WIP, so there is indeed missing support currently for the `dim` argument.

**Comment by sjain-stanford** [comment]
For the record, I plan to resurrect https://github.com/llvm/torch-mlir/pull/1063 since this effort / PR seems to be stale now. Please feel free to push back if that's not true..

**Comment by ramiro050** [comment]
Yes, I think we can consider this stale, since there hasn't been an update in a year and a half.

**Comment by riestmo-nxp** [comment]
Are your plans to resurrect https://github.com/llvm/torch-mlir/pull/1063 still active @sjain-stanford? I'm currently blocked by this error:

error: failed to legalize operation 'torch.operator' that was explicitly marked illegal
    %204 = torch.operator "torch.aten.index_add"(%165, %int0_249, %182, %203, %int1_250) : (!torch.vtensor<[3,2048],f32>, !torch.int, !torch.vtensor<[?],si64>, !torch.vtensor<[?,2048],f32>, !torch.int) -> !torch.vtensor<[3,2048],f32>

 Therefore I would be happy to see support for aten::index_add added.
